universal-application-tool-0.0.1/app/repository/UserRepository.java

```diff
@@ -55,40 +57,54 @@ public UserRepository(
-   * Returns all programs that are appropriate to serve to an applicant - which is any active
-   * program, plus any program where they have an application in the draft stage.
+   * Returns all programs that are appropriate to serve to an applicant - which is any program
+   * program where they have an application in the draft stage, and any active program. Programs
+   * with an application in the draft stage and active programs are deduped, with the ones with an
+   * application in the draft stage taking priority.
    */
-  public CompletionStage<ImmutableList<ProgramDefinition>> programsForApplicant(long applicantId) {
+  public CompletionStage<ImmutableMap<LifecycleStage, ImmutableList<ProgramDefinition>>>
+      programsForApplicant(long applicantId) {
     return supplyAsync(
-            () -> {
-              List<Program> inProgressPrograms =
-                  ebeanServer
-                      .find(Program.class)
-                      .alias("p")
-                      .where()
-                      .exists(
-                          ebeanServer
-                              .find(Application.class)
-                              .where()
-                              .eq("applicant.id", applicantId)
-                              .eq("lifecycle_stage", LifecycleStage.DRAFT)
-                              .raw("program.id = p.id")
-                              .query())
-                      .endOr()
-                      .findList();
-              List<Program> activePrograms =
-                  versionRepositoryProvider.get().getActiveVersion().getPrograms();
-              return new ImmutableList.Builder<Program>()
-                  .addAll(activePrograms)
-                  .addAll(inProgressPrograms)
-                  .build();
-            },
-            executionContext.current())
-        .thenApplyAsync(
-            (programs) ->
-                programs.stream()
-                    .map(program -> program.getProgramDefinition())
-                    .collect(ImmutableList.toImmutableList()));
+        () -> {
+          List<Program> inProgressPrograms =
+              ebeanServer
+                  .find(Program.class)
+                  .alias("p")
+                  .where()
+                  .exists(
+                      ebeanServer
+                          .find(Application.class)
+                          .where()
+                          .eq("applicant.id", applicantId)
+                          .eq("lifecycle_stage", LifecycleStage.DRAFT)
+                          .raw("program.id = p.id")
+                          .query())
+                  .endOr()
+                  .findList();
+          Set<String> draftProgramAdminNames =
+              inProgressPrograms.stream()
+                  .map(program -> program.getProgramDefinition().adminName())
+                  .collect(Collectors.toSet());
+
+          // Get active programs, but dedupe the ones with a draft application.
```

Comment:
Suggestion (non-blocking): Since you are returned a map, you may delay deduping to the consumers of this function.

Reply:
that's a great idea!

Reply:
did the change!